src/Stack/Config.hs

```diff
@@ -590,19 +599,51 @@ resolvePackageLocation menv projRoot (PLRemote url remotePackageType) = do
 
         _ -> return dir
 
--- | Get the stack root, e.g. ~/.stack
-determineStackRoot :: (MonadIO m, MonadThrow m) => m (Path Abs Dir)
-determineStackRoot = do
-    env <- liftIO getEnvironment
-    case lookup stackRootEnvVar env of
-        Nothing -> do
-            x <- liftIO $ getAppUserDataDirectory stackProgName
-            parseAbsDir x
-        Just x -> do
-            y <- liftIO $ do
-                createDirectoryIfMissing True x
-                canonicalizePath x
-            parseAbsDir y
+-- | Get the stack root, e.g. ~/.stack, and determine whether the user owns it.
+--
+-- On Windows, the second value is always 'True'.
+determineStackRootAndOwnership :: (MonadIO m, MonadThrow m) => m (Path Abs Dir, Bool)
+determineStackRootAndOwnership = do
+    stackRoot <- liftIO $ do
+        mstackRoot <- lookupEnv stackRootEnvVar
+        case mstackRoot of
+            Nothing -> getAppUserDataDirectory stackProgName
+            Just x -> return x
+    stackRootExists <- liftIO $ doesDirectoryExist stackRoot
+
+    -- This allows us to avoid a lot of CPP in the imports section.
+#ifdef WINDOWS
+    let osIsWindows = True
+#else
+    let osIsWindows = False
+#endif
+
+    userOwnsStackRootOrParentDir <-
+        if osIsWindows
+            then return True
+            else do
+                existingStackRootOrParentDir <- do
```

Comment:
> Can you please wrap up this "search parent directories till a predicate is met" into a utility in Path.IO?

Will do. There's something wrong with your link though. Is this the spot you were referring to?
https://github.com/commercialhaskell/stack/blob/0347cd15c3fbb818e0877e0b267d0af4d121f66d/src/Stack/Config.hs#L674

Regarding the type, we'd need to allow monadic predicates, so that would be `MonadIO m => (Path Abs Dir -> m (Maybe a)) -> Path Abs Dir -> m (Maybe a)`.



> It'd best to avoid this nubOrd / iterate pattern, because it cannot leverage laziness, and will always compute the 100 intermediate paths.

I'm not quite sure I understand this correctly but wouldn't the following result in an infinite loop if it were true?
```
λ> let xs = nubOrd $ iterate succ (1 :: Int)
λ> head xs
1
λ> :sprint xs
xs = 1 : _
```

But terminating when we the directory is the same as the previous one, as the other implementation does, is much better anyways.